openshift-knative/serverless-operator
# Add upstream Eventing dashboards

- Adds Eventing dashboards for broker/trigger and pingsource/apiserver source directly from the upstream one.
- Uses two separate files for the eventing dashboards since Openshift monitoring does not apply multiple data in a config map.
- Creates the required service monitors and services to get the metrics from the Eventing pods.
Resources created depend on the current Eventing instance but they are applied only if they dont exist.
- Targets in Prometheus stay empty when no pod is available eg. pingsource adapter:
![image](https://user-images.githubusercontent.com/7945591/93656820-e0cd8f00-fa35-11ea-87bb-bc8953c9b39d.png)
When the user creates a pingsource then the adapter pod is created and then target is updated.
- Fixes several minor issues wrt monitoring
- Adds back the CR metrics endpoint which can be expanded further in the future, currently we support the two basic CRD types
![image](https://user-images.githubusercontent.com/7945591/93746371-ab739d80-fbfd-11ea-944b-f8c3a1022649.png)
- Adds an e2e test for checking components presence eg. operator's service monitor
- Adds a controller to monitor namespaces for apiserver source creation or in general eventing source deployments and then create the required service monitors. Right now user needs to manually create them according to the yaml file above.

This PR was tested with crc 4.5.9.

To test the dashboards for trigger/filter and pingsource apply the following contents in ns `default`:
```
apiVersion: eventing.knative.dev/v1beta1
kind: Trigger
metadata:
  annotations:
    knative-eventing-injection: enabled
  name: example-trigger
spec:
  broker: default
  subscriber: 
    ref:
      apiVersion: serving.knative.dev/v1
      kind: Service
      name: event-display 
---
apiVersion: sources.knative.dev/v1alpha2
kind: PingSource
metadata:
  name: ping-source
spec:
  schedule: "*/1 * * * *"
  jsonData: '{"message": "Hello serverless!"}'
  sink:
    ref:
      apiVersion: eventing.knative.dev/v1beta1
      kind: Broker
      name: default
---
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: event-display
spec:
  template:
    spec:
      containers:
        - image: quay.io/openshift-knative/knative-eventing-sources-event-display
```
To test the apiserver source run:
```
kubectl create ns apisource
kubectl label ns apisource openshift.io/cluster-monitoring=true
```
and then apply in ns `apisource` this [yaml](https://gist.github.com/skonto/3f89171b701553db0df62cb5d86579a2).

Screenshots of the dashboards:
![image](https://user-images.githubusercontent.com/7945591/93656966-1626ac80-fa37-11ea-819a-96de5bca2c2d.png)
![image](https://user-images.githubusercontent.com/7945591/93656971-1cb52400-fa37-11ea-9feb-6b4388490b7d.png)
![image](https://user-images.githubusercontent.com/7945591/93656974-2474c880-fa37-11ea-9477-098ce34665ff.png)
![image](https://user-images.githubusercontent.com/7945591/93656977-2a6aa980-fa37-11ea-8e3c-0874889eb287.png)
![image](https://user-images.githubusercontent.com/7945591/93656982-30608a80-fa37-11ea-9a24-3e686039c5bd.png)
![image](https://user-images.githubusercontent.com/7945591/93657036-9f3de380-fa37-11ea-8650-204a81bfaa25.png)
![image](https://user-images.githubusercontent.com/7945591/93656990-39e9f280-fa37-11ea-97ac-aae19e667d46.png)


Note: a lot of files are vendor files needed to pass `dep check` and bring the right deps for the new e2e test.
